### PR TITLE
core models and synchronized methods

### PR DESCRIPTION
**Pull request by dcattaruzza**
Features relating to Atomic sections, monitors, synchronized methods and core models for java integrated in the cbmc executable
Also fixed several errors in the exception handling mechanisms

### CODE REVIEW

File: src/java_bytecode/java_bytecode_synchronize.cpp [hunk 1]
```diff
@@ -135,10 +135,10 @@ Author: Daniel Kroening, kroening@kroening.com
     code_function_callt call;
     if (is_enter)
       call.function()=
-        symbol_exprt("java::java.lang.Object.monitorenter:()V", type);
+        symbol_exprt("java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V", type);
```
Comment (cesaro) [inline src/java_bytecode/java_bytecode_synchronize.cpp]: Change file author

File: src/java_bytecode/library/src/java/lang/Object.java [hunk 2]
```diff
@@ -147,25 +147,30 @@ public final void wait() throws InterruptedException {
 
     protected void finalize() throws Throwable { }
     
-    public void monitorenter()
+    public static void monitorenter(Object object)
```
Comment (cesaro) [inline src/java_bytecode/library/src/java/lang/Object.java:150]: Check entire file for proper indentation.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 3]
```diff
@@ -2693,9 +2693,7 @@ codet java_bytecode_convert_methodt::convert_instructions(
     for(const auto &exception_row : method.exception_table)
     {
       // add the CATCH-POP before the end of the try block
-      if(cur_pc<exception_row.end_pc &&
-         !working_set.empty() &&
-         *working_set.begin()==exception_row.end_pc)
+      if(cur_pc==exception_row.end_pc)
```
Comment (cesaro) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Better commit explanation.

File: regression/cbmc-java/exceptions27/test.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      // Account for the rethrow in finally
+      j++;
+    }
+    catch(B exc)
+    {
+      i++;
+    }
+    // Make sure B was rethrown by finally when A was thrown
+    // or if B was thrown there was no rethrow
+//    assert(i+j == 4);*/
```
Comment (cesaro) [inline regression/cbmc-java/exceptions27/test.java]: indentation

File: src/java_bytecode/jar_file.cpp [hunk 5]
```diff
@@ -26,6 +26,21 @@ jar_filet::jar_filet(
   }
 }
 
+jar_filet::jar_filet(
```
Comment (cesaro) [inline src/java_bytecode/jar_file.cpp]: Reword commit title to "Add constructor to be able to load in-memory JAR files"

File: src/java_bytecode/jar_file.cpp [hunk 6]
```diff
@@ -26,6 +26,21 @@ jar_filet::jar_filet(
...
+  java_class_loader_limitt &limit,
+  const void *pMem,
+  size_t size):
+  m_zip_archive(pMem,size)
+{
```
Comment (cesaro) [inline src/java_bytecode/jar_file.cpp]: Refaactor this into a method `initialize_file_index`

File: src/java_bytecode/jar_file.h [hunk 7]
```diff
@@ -26,6 +26,14 @@ class jar_filet final
   /// \param filename Name of the file
   /// \throw Throws std::runtime_error if file cannot be opened
   jar_filet(java_class_loader_limitt &limit, const std::string &filename);
+
+  /// Open java buffer for reading
```
Comment (cesaro) [inline src/java_bytecode/jar_file.h]: Suggestion: `Open a JAR file of size \p size loaded in memory at address \p pMem`

File: src/java_bytecode/Makefile [hunk 8]
```diff
@@ -29,16 +29,23 @@ SRC = bytecode_info.cpp \
...
+      java_core_models.c \
       # Empty last line
 
 INCLUDES= -I ..
 
 include ../config.inc
 include ../common
 
-CLEANFILES = java_bytecode$(LIBEXT)
+CLEANFILES = java_bytecode$(LIBEXT) java_core_models.c java_core_models.h
```
Comment (cesaro) [inline src/java_bytecode/Makefile]: Title: core-models -> core-models.jar

File: src/java_bytecode/Makefile [hunk 9]
```diff
@@ -72,7 +72,7 @@ converter: library/converter.cpp
 	$(CXX) $(CP_CXXFLAGS) -o $@ $<
 
 java_class_loader.o: java_core_models.inc java_class_loader.cpp
-	$(CXX) -c $(CP_CXXFLAGS) -o $@ $<
+	$(CXX) -c $(CP_CXXFLAGS) -o $@ java_class_loader.cpp
```
Comment (cesaro) [inline src/java_bytecode/Makefile]: Commit can be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/ansi-c/Makefile [hunk 10]
```diff
@@ -121,7 +121,7 @@ ansi_c_internal_additions$(OBJEXT): arm_builtin_headers.inc
 ansi_c_internal_additions$(OBJEXT): cw_builtin_headers.inc
 ansi_c_internal_additions$(OBJEXT): clang_builtin_headers.inc
 
-generated_files: cprover_library.inc gcc_builtin_headers_generic.inc \
+generated_files: c#prover_library.inc gcc_builtin_headers_generic.inc \
```
Comment (cesaro) [inline src/ansi-c/Makefile]: Error!

File: src/java_bytecode/library/Makefile [hunk 11]
```diff
@@ -1,32 +0,0 @@
-all: core-models.jar converter
-SOURCE_DIR := src
-BINARY_DIR := classes
-
-FIND := find
```
Comment (cesaro) [inline src/java_bytecode/library/Makefile]: Instead of removing it, do:

```
all :
    make -C .. core-models.jar
```

File: .gitignore [hunk 12]
```diff
@@ -38,6 +38,7 @@ src/ansi-c/gcc_builtin_headers_tm.inc
 src/ansi-c/gcc_builtin_headers_mips.inc
 src/ansi-c/gcc_builtin_headers_power.inc
 src/ansi-c/gcc_builtin_headers_ubsan.inc
+src/java_bytecode/java_core_models.inc
```
Comment (cesaro) [inline .gitignore:41]: Commit can also be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/java_bytecode/java_class_loader.cpp [hunk 13]
```diff
@@ -125,19 +125,22 @@ java_bytecode_parse_treet &java_class_loadert::get_parse_tree(
...
-		   core_models,
-		   java_core_models,
-		   java_core_models_size);
-
-  // This does not read from the jar file but from the jar_filet object
-  // as we've just created it
-  read_jar_file(class_loader_limit, core_models);
-  if (get_class_file(class_loader_limit, class_name, core_models, parse_tree))
-    return parse_tree;
+  if (prefer_core_models)
```
Comment (cesaro) [inline src/java_bytecode/java_class_loader.cpp]: Refactor this hunk and next hunk into a function.

File: regression/cbmc-java/destructor1/test.desc [hunk 14]
```diff
@@ -1,6 +1,6 @@
 CORE
 Break.class
---show-goto-functions
+--show-goto-functions --lazy-methods
```
Comment (cesaro) [inline regression/cbmc-java/destructor1/test.desc:3]: Merge this commit with previous one

File: src/java_bytecode/java_bytecode_language.cpp [hunk 15]
```diff
@@ -507,11 +508,12 @@ void java_bytecode_languaget::convert_threadblock(codet &code,
     code_labelt tmp_c(lbl1);
     tmp_c.op0()=codet(ID_skip);
 
-    exprt plus(ID_plus, java_int_type());
-    plus.copy_to_operands(next_symbol.symbol_expr());
-    plus.copy_to_operands(from_integer(1, java_int_type()));
-    code_assignt tmp_d(current_symbol.symbol_expr(), next_symbol.symbol_expr());
-    code_assignt tmp_e(next_symbol.symbol_expr(), plus);
+      exprt plus(ID_plus, java_int_type());
```
Comment (cesaro) [inline src/java_bytecode/java_bytecode_language.cpp]: Spacing problem here?

File: src/java_bytecode/Makefile [hunk 16]
```diff
@@ -39,15 +39,17 @@ include ../common
 
 CLEANFILES = java_bytecode$(LIBEXT) java_core_models.c java_core_models.h
 
-all: subsystem run java_bytecode$(LIBEXT)
+all: subsystem run core-models java_bytecode$(LIBEXT)
```
Comment (cesaro) [inline src/java_bytecode/Makefile]: Commit can be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/java_bytecode/java_class_loader.h [hunk 17]
```diff
@@ -23,6 +23,13 @@ Author: Daniel Kroening, kroening@kroening.com
 class java_class_loadert:public messaget
 {
 public:
+  // Default constructor does not use core models
+  // for backward compatibility of unit tests
```
Comment (cesaro) [inline src/java_bytecode/java_class_loader.h:27]: 1. Try to see if the unit tests pass now when the two fields are set to true
2. If not, add a FIXME here saying that the two fields should be true and the unit tests should be fixed.

File: src/java_bytecode/CMakeLists.txt [hunk 18]
```diff
@@ -18,9 +18,11 @@ add_custom_command(TARGET core_models_files
 file(GLOB_RECURSE c_sources "*.c")
 file(GLOB_RECURSE sources "*.cpp")
 file(GLOB_RECURSE headers "*.h")
+set_source_files_properties("java_core_models.inc" GENERATED)
```
Comment (cesaro) [inline src/java_bytecode/CMakeLists.txt]: Commit can be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/java_bytecode/CMakeLists.txt [hunk 19]
```diff
@@ -1,6 +1,26 @@
+get_filename_component(JAVA_CORE_LIBRARY_PATH "library" REALPATH BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+get_filename_component(CORE_MODELS_JAR "library/core-models.jar" REALPATH BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (cesaro) [inline src/java_bytecode/CMakeLists.txt]: Commit can be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/java_bytecode/Makefile [hunk 20]
```diff
@@ -66,11 +66,14 @@ core-models.jar: $(all_classes)
 
 CLEANFILES = java_bytecode$(LIBEXT) core-models.jar converter all_classes
 
-all: | java_core_models.inc java_bytecode$(LIBEXT)
+all: java_bytecode$(LIBEXT)
```
Comment (cesaro) [inline src/java_bytecode/Makefile]: Commit can be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/java_bytecode/CMakeLists.txt [hunk 21]
```diff
@@ -15,8 +15,7 @@ add_custom_command(TARGET core_models_files
                    PRE_BUILD
                    COMMAND java-converter ${CORE_MODELS_JAR} "java_core_models" ${CMAKE_CURRENT_SOURCE_DIR})
 
-file(GLOB_RECURSE c_sources "*.c")
-file(GLOB_RECURSE sources "*.cpp")
+file(GLOB sources "*.cpp")
```
Comment (cesaro) [inline src/java_bytecode/CMakeLists.txt]: Commit can be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/java_bytecode/CMakeLists.txt [hunk 22]
```diff
@@ -1,19 +1,20 @@
-get_filename_component(JAVA_CORE_LIBRARY_PATH "library" REALPATH BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+find_package(Java REQUIRED)
+include(UseJava)
```
Comment (cesaro) [inline src/java_bytecode/CMakeLists.txt:2]: Commit can be squashed into https://github.com/cesaro/cbmc/pull/5/commits/068c730d5164434a1b3c5894d57b04e361cef962

File: src/java_bytecode/java_class_loader.h (hunk 17 above)
Comment (dcattaruzza) [reply src/java_bytecode/java_class_loader.h:27]: Changed to default use_core_models=true. Tests pass

**Pull request merged**